packages/core/http/core-http-server/src/router/scratch_pad.ts

```diff
@@ -0,0 +1,139 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { schema } from '@kbn/config-schema';
+import type { Logger } from '@kbn/logging';
+import { pipe } from 'fp-ts/lib/function';
+import type { RequestHandler } from './request_handler';
+import type { RequestHandlerContextBase } from './request_handler_context';
+import type { RouteConfig, RouteMethod } from './route';
+import type { IRouter, RouteRegistrar } from './router';
+import type { RouteValidatorFullConfig } from './route_validator';
+
+const logger: Logger = {} as any;
+/** Definitely subject to revision, just used for examples */
+type Version = '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10';
+
+/**
+ * ===================== Fourth design =====================
+ * Take the lib/toolkit approach. Instead of baking the idea of versioning into
+ * the router we can expose a toolkit for building a versioned API that uses our
+ * HTTP router. The toolkit just wraps the router ensuring that the spec for
+ * applying/extracting version information is followed.
+ */
+
+{
+  /**
+   * We must omit "validate" because we are going to declare multiple validations per route.
+   */
+  type VersionedRouteOpts = Omit<RouteConfig<unknown, unknown, unknown, RouteMethod>, 'validate'>;
+  /**
+   * Simplify the Registrar type by pre assigning the RouteMethod generic type
+   */
+  type Registrar<Context extends RequestHandlerContextBase> = RouteRegistrar<RouteMethod, Context>;
+  /**
+   * This is the primary interface for the toolkit
+   */
+  interface VersionedAPIToolkit {
+    defineRoute<Context extends RequestHandlerContextBase>(
+      registrar: Registrar<Context>,
+      opts: VersionedRouteOpts
+    ): VersionedRoute<Context>;
+  }
+  /**
+   * The toolkit defines versioned routes, taking care of all the versioning shenanigans
+   * in the background.
+   */
+  interface VersionedRoute<Context extends RequestHandlerContextBase = RequestHandlerContextBase> {
+    addVersion<P, Q, B>(
+      version: Version,
+      opts: { validation: RouteValidatorFullConfig<P, Q, B> },
+      handler: RequestHandler<P, Q, B, Context>
+    ): VersionedRoute<Context>;
+  }
+
+  const vtk: VersionedAPIToolkit = {} as any;
+  /** A router with some custom context */
+  const myRouter: IRouter<{ test: number } & RequestHandlerContextBase> = {} as any;
+
+  const versionedRoute = vtk
+    .defineRoute(myRouter.post, { path: '/api/my-plugin/my-route', options: {} })
+    .addVersion(
+      '1',
+      {
+        validation: { body: schema.object({ n: schema.number({ min: 0, max: 1 }) }) },
+      },
+      async (ctx, req, res) => {
+        logger.info(String(ctx.test));
+        logger.info(String(req.body.n));
+        return res.ok();
+      }
+    )
```

Comment:
This is very close to the design we came with for the initial versioned API RFC we presented last year.

The main difference is that my design was more tightly coupling the (unversioned) router and the versioned one (as the versioning API was accessible via something like `core.http.createRouter().versioned.createRoute`.

Overall I think fully separating the two is more elegant. 

I would still avoid passing the router for each route definition.

Something like

```ts
const versionedApi = vtk.createVersionedAPI(router);
const versionedRoute = versionedApi
    .defineRoute({ method: 'POST', path: '...', blabla })
    .addVersion(same params as your example);

// fluent API ok, but also usable later
versionedRoute.addVersion(...)
```

(Still need to figure out what the best way to propagate/specify the method is, probably)

Note that the main risk I see with this approach is that we will be 'wrapping' our http router as any other non-Core consumer would, meaning that any capabilities required by the versioning API will have to be exposed publicly on our router. (I don't think this will be an issue, but it's definitely worth mentioning)

Reply:
I like Pierre's suggestions :+1:

I would use named params for `addVersion` similar to `defineRoute`

```
const versionedRoute = versionedApi .defineRoute({ method: 'POST', path: '...', blabla }) .addVersion({
  version: '1',
  validation: {...}, // Do we need a bag of "options" or could we just add them as top-level named params like I did here?
  anotherOptionWeMightAddLater: '',
  handler: () => {},
});
```

Reply:
Another option is to something like
```
const casesRouter = core.http.createRouter('/api/cases');
const casesV1Router = vtk.createVersionedRouter(casesRouter, 'v1')
casesV1Router.get( /*api identical to what we have today*/)
```

Not sure I prefer it... it makes it easier to see everything that's included in V1 of the API, but harder to compare implementations between versions of the same endpoint. Probably the latter is more useful.

So I think I prefer starting with the route and then adding all versions to it instead of this example where we start with the version and then add routes. But thought I'll add it to the list of options.

Reply:
Yeah, it's definitely another option.

My feeling is that it's most common to group per endpoint ( endpoint1 -> v1/v2/3 ) than per version ( v1 -> endpoint 1/2/3). I could be wrong though.